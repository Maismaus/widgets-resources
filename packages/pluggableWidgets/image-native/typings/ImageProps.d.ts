/**
 * This file was generated from ImageViewer.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix UI Content Team
 */
import { ActionValue, DynamicValue, NativeIcon, NativeImage } from "mendix";

export type DatasourceEnum = "image" | "imageUrl" | "icon";

export type OnClickTypeEnum = "action" | "enlarge";

export interface ImageViewerProps<Style> {
    name: string;
    style: Style[];
    datasource: DatasourceEnum;
    imageObject?: DynamicValue<NativeImage>;
    defaultImageDynamic?: DynamicValue<NativeImage>;
    imageUrl?: DynamicValue<string>;
    imageIcon?: DynamicValue<NativeIcon>;
    onClickType: OnClickTypeEnum;
    onClick?: ActionValue;
}

export interface ImageViewerPreviewProps {
    class: string;
    style: string;
    datasource: DatasourceEnum;
    imageObject: { type: "static"; imageUrl: string } | { type: "dynamic"; entity: string } | null;
    defaultImageDynamic: { type: "static"; imageUrl: string } | { type: "dynamic"; entity: string } | null;
    imageUrl: string;
    imageIcon: { type: "glyph"; iconClass: string } | { type: "image"; imageUrl: string } | null;
    onClickType: OnClickTypeEnum;
    onClick: {} | null;
}
