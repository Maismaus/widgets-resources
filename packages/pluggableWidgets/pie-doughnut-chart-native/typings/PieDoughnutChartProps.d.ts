/**
 * This file was generated from PieDoughnutChart.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix UI Content Team
 */
import { ListValue, ListAttributeValue } from "mendix";
import { Big } from "big.js";

export type PresentationEnum = "pie" | "doughnut";

export interface SeriesType {
    dataSource: ListValue;
    XAttribute: ListAttributeValue<Date | string>;
    YAttribute: ListAttributeValue<Big>;
    sliceStylingKey?: ListAttributeValue<string>;
}

export type SortOrderEnum = "ascending" | "descending";

export interface SeriesPreviewType {
    dataSource: {} | { type: string } | null;
    XAttribute: string;
    YAttribute: string;
    sliceStylingKey: string;
}

export interface PieDoughnutChartProps<Style> {
    name: string;
    style: Style[];
    presentation: PresentationEnum;
    series: SeriesType[];
    sortOrder: SortOrderEnum;
    showLabels: boolean;
}

export interface PieDoughnutChartPreviewProps {
    class: string;
    style: string;
    presentation: PresentationEnum;
    series: SeriesPreviewType[];
    sortOrder: SortOrderEnum;
    showLabels: boolean;
}
