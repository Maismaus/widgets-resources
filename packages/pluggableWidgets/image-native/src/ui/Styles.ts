import { ViewStyle, ImageStyle } from "react-native";
import { Style } from "@mendix/piw-native-utils-internal";

export interface CustomImageStyle extends ImageStyle {
    svgColor?: string;
}

export interface ImageViewerStyle extends Style {
    container: ViewStyle;
    image: CustomImageStyle;
    backdrop: ViewStyle;
}

export const defaultImageViewerStyle: ImageViewerStyle = {
    container: {},
    image: {},
    backdrop: {
        width: "100%",
        height: "100%",
        justifyContent: "center",
        alignItems: "center",
        backgroundColor: `rgba(0,0,0,0.8)`
    }
};
