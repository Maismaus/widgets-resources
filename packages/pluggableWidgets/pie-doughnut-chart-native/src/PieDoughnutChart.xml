<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.native.piedoughnutchart.PieDoughnutChart" supportedPlatform="Native" offlineCapable="true"
        pluginWidget="true" xmlns="http://www.mendix.com/widget/1.0/"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../../node_modules/mendix/custom_widget.xsd">
    <name>Pie Doughnut Chart</name>
    <description>Render a series of data in a Pie or Doughnut chart.</description>
    <icon>iVBORw0KGgoAAAANSUhEUgAAAEAAAABACAYAAACqaXHeAAAACXBIWXMAAAsTAAALEwEAmpwYAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAXMSURBVHgB7Zo7U+NWFIDPlY2Ds8nE+QfahgKGh7t0EUUypIo1PDvYX7DQpcOuUga6dPF2DGDsrcJkC0SXzt4Chm61vyDeycMkYCvnxDIjXV9ZD18Zmfib0Xh1JS26555z7nkIYMKECf9nGCSElVrLsv9p0mEx1oBO530HOsYb/ZMGxEQSBSDCxMNoW+1D2cIYFwE8wIA1LMs6PNezZZDA2AnAgQkWLKMgTBgCBRIKrvIeA6tmgdX0uEXF5Xv3zeu/ftCqVg4iklgNOC9kH97t6+qfhRRj2xawgsfjkbUhsRrg5Bf9We3nwsc6TvI5WJ1XgltUxlj9q+ofSxCSsRBAD1rhc/3ZTttK5aG7MzyAppJLsVR95fU/2xCCWEygUCjkpqamCmjHi4qiqPhLK0N2mjs5ORH+zUEmIEKr/pbLwkdFjBde8tdwu8wH3S7TIAmadCaTeYmT1ej9aAzVkpwZxIGhf07OcXel2mriMu47r6VZ+gLH80F8wtAa4Jj4LnRXORC8JoTVANez1VaRFwJi3lrTeUNnzUHPDuUDNjY29lHV3+HkixBi8rLBlS6iEyhxw2pWae37PRtJAFtbW+r6+np90MTxmtHpdA7x0O/u7p5DzJAQGEaI7ndgaCK/a4OeC+0DcNW32+32AYgn3qQw9f7+/qBWq7lUDwUGcdOCv4vTMP0tUJDUg6VJCwyvZ0IJgFTeXnUemnjp9PT0AB4RcozoD16gP7hwDGukBef6p4bomcAm4DV5UnNS8ceefA80BYNCaNdgVwuEBBLA6upqQTR5itcrlcour+6Pzb2V5h2i5uULfAVADg+DmZ+4YVL55aSsOs8bPUNBkOEcsyAlzCN8BYAOj+yJd3g0eQOSDLctKkwRhsgDBUB2D06PCl21xyAmthKVLG7htuFMpSlXECVLngIg1RfYfTmpas/TDZUVwzmWBkXj7/MUAKp+X2iJ3r4E44TVuXSdMrbI3yKMA2j1UQA73HAJvb0JIyJCicwXNINgJoCT17ghE+2+DGMOFlT7olehAND2+Rx7vFTfG5Uf6BMAqT/m8S5VQds34InS5wN49aesbhS2Hyb/l0mfBmBszzuKS3jC9AkAw17XVoEakPigZxhETlB13aAoJjxhRHGA6jxBB2hCDGCBRPY+b6LzLh4fH78K85BvMpS0VHcAFLqHDtNH2RgZhSBDF2ZFAnC9KJW9QQ57wHVzkoDIB5AAHiaNTvAzkLB6dihdBomgH6Etu+4YMiEkIg1wbXvoWPKQUDBmUZ3n6ANMCIkoEHrPDWmQUPiQHQXwFkIiCoQa3PkiJBQUwJfOc/7dg9AnANz3a9yQJtERSoOSNuC0M0rS1icAe983nGPpdHoXEoagZtGIkrR51QNcCRAT9OATAF+yO4QICAVAvT1wb325zc3NxGgBbn870B+yGxABoQDIDCyu04q7w34SfIFt+/zql6PWLDxDYZEWZDIZ33573NjVatU5Nky1OuV14ebm5nZ2djaL9q85hr+Yn5//cHV19Ss8AmiG9CXKd84x6kqfnZ3VICK+ZSj6EAJ/+IBj5K0xQdhLULV6qI8vfLPBVCqlA5cLoFZU7RcaCWtraxr+XHDDTVT9ZRgSXwEcHR2Z0M3knJAzrI9iZyC1R4H3NWjRKb+QUaxNBbnp+vq6gf6Acf6ATGFlbm5OnZmZeYs+Q2q+TzvOwsLC9x7fJZQqlcqPIIFAAiBQCIZICMgSmkkBBfGBBAUSoFXHuJ4cm8Zfsz/FKYIkQtfi6WsR+4MJUUxg4lGioCSsegb43pA0bE92iy5SM8JunpJdqgNuMyikprI6VZapuNqrL9oBVQ41Z4kyOjut1Qb8Xw18Xo+jQTNUNwa9cxFfPs7g6L+IVKbK8wzdjiJtwKiRBBHqK20fPL83lI20fpxtFhp1lvlKTVCoD4k/l6OYeI9YGpI9YVCfEYWxiIcKbn/RtI8GleCokkOFmDHqQUyYMOGJ8C9UisDhnfIOIQAAAABJRU5ErkJggg==</icon>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="Presentation">
                <property key="presentation" type="enumeration" defaultValue="pie">
                    <caption>Presentation</caption>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="pie">Pie</enumerationValue>
                        <enumerationValue key="doughnut">Doughnut</enumerationValue>
                    </enumerationValues>
                </property>
            </propertyGroup>
            <propertyGroup caption="Data source">
                <property key="series" type="object" isList="true">
                    <caption>Data series</caption>
                    <description>Add one or more data series.</description>
                    <properties>
                        <propertyGroup caption="General">
                            <property key="dataSource" type="datasource" isList="true">
                                <caption>Data source</caption>
                                <description>Data points for individual series.</description>
                            </property>
                            <property key="XAttribute" type="attribute"
                                      dataSource="dataSource">
                                <caption>X attribute</caption>
                                <description/>
                                <attributeTypes>
                                    <attributeType name="DateTime"/>
                                    <attributeType name="String"/>
                                    <attributeType name="Enum"/>
                                </attributeTypes>
                            </property>
                            <property key="YAttribute" type="attribute"
                                      dataSource="dataSource">
                                <caption>Y attribute</caption>
                                <description/>
                                <attributeTypes>
                                    <attributeType name="Decimal"/>
                                    <attributeType name="Integer"/>
                                    <attributeType name="Long"/>
                                </attributeTypes>
                            </property>
                            <property key="sliceStylingKey" type="attribute" dataSource="dataSource" required="false">
                                <caption>Dynamic styling key</caption>
                                <description>Configure which attribute is used as the key that matches with styling defined in Atlas.</description>
                                <attributeTypes>
                                    <attributeType name="String"/>
                                    <attributeType name="Enum"/>
                                </attributeTypes>

                            </property>
                        </propertyGroup>
                    </properties>
                </property>
                <property key="sortOrder" type="enumeration" defaultValue="ascending">
                    <caption>Sort order</caption>
                    <description>Applicable only to Y attribute values.</description>
                    <enumerationValues>
                        <enumerationValue key="ascending">Ascending</enumerationValue>
                        <enumerationValue key="descending">Descending</enumerationValue>
                    </enumerationValues>
                </property>
            </propertyGroup>
            <propertyGroup caption="Chart settings">
                <property key="showLabels" type="boolean" defaultValue="true">
                    <caption>Show labels</caption>
                    <description/>
                </property>
            </propertyGroup>
            <propertyGroup caption="Common">
                <systemProperty key="Name"/>
                <systemProperty key="Visibility"/>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
