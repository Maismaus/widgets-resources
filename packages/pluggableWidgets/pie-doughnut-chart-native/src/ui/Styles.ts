import { ViewStyle } from "react-native";

interface LabelStyle {
    // color is the same as slice color
    color?: string;
    fontFamily?: string;
    fontSize?: number;
    fontStyle?: "normal" | "italic";
    fontWeight?: "normal" | "bold" | "100" | "200" | "300" | "400" | "500" | "600" | "700" | "800" | "900";
}

interface SliceStyle {
    color?: string;
}

export interface SlicesStyle {
    customStyles?: {
        [key: string]: {
            slice?: SliceStyle;
            label?: LabelStyle;
        };
    };
    colorPalette?: string;
    innerRadius?: number;
    padding?: number;
    paddingBottom?: number;
    paddingHorizontal?: number;
    paddingLeft?: number;
    paddingRight?: number;
    paddingTop?: number;
    paddingVertical?: number;
}

export interface ChartStyle {
    container?: ViewStyle;
    slices?: SlicesStyle;
}

export const defaultStyle: ChartStyle = {
    container: {
        flex: 1
    },
    slices: {
        padding: 40
    }
};
