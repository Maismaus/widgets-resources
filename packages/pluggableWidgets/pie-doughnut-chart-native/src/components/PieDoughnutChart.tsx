import { createElement, ReactElement, useCallback, useMemo, useState } from "react";
import { VictoryPie } from "victory-native";
import { VictoryStyleObject, CallbackArgs, BlockProps } from "victory-core";

import { ChartStyle, SlicesStyle } from "../ui/Styles";
import { SortOrderEnum } from "../../typings/PieDoughnutChartProps";
import { Option } from "mendix";
import { LayoutChangeEvent, View } from "react-native";

const DEFAULT_INNER_RADIUS_RATIO = 12;
const DEFAULT_SLICE_COLOUR = "black";
const LABEL_STYLE_KEYS = ["fill", "fontFamily", "fontSize", "fontStyle", "fontWeight"];

export interface ChartProps {
    series: DataPoints;
    sortOrder: SortOrderEnum;
    style: ChartStyle;
    presentation: string;
    showLabels: boolean;
}

export type DataPoints = Array<Slice<string, number>>;

export interface Slice<X extends string, Y extends number> {
    x: X;
    y: Y;
    stylingKey: Option<string>;
}

export function PieDoughnutChart({
    presentation,
    series,
    sortOrder,
    style,
    showLabels
}: ChartProps): ReactElement | null {
    // due to the nature of the chart type, we only reply on the width, as the chart is always a square
    const [chartDimensions, setChartDimensions] = useState<{ width: number }>();
    // Chart user-styling may be missing for certain slices. A palette is passed, any missing colours
    // fallback to a colour from the palette or the default color.
    const normalizedSliceColors: string[] = useMemo(() => {
        const sliceColorPalette = style.slices?.colorPalette?.split(";");
        let index = 0;

        return series.map(_series => {
            const configuredStyle = !_series.stylingKey
                ? null
                : style.slices?.customStyles?.[_series.stylingKey]?.slice?.color;
            if (typeof configuredStyle !== "string") {
                const color = sliceColorPalette?.[index] || DEFAULT_SLICE_COLOUR;
                if (sliceColorPalette) {
                    index = index + 1 === sliceColorPalette.length ? 0 : index + 1;
                }

                return color;
            }

            return configuredStyle;
        });
    }, [series, style]);

    const updateChartDimensions = useCallback(
        (event: LayoutChangeEvent) =>
            setChartDimensions({
                width: event.nativeEvent.layout.width
            }),
        [setChartDimensions]
    );

    return (
        <View style={{ ...style.container }}>
            <View
                onLayout={updateChartDimensions}
                style={{ justifyContent: "center", alignItems: "center" }}
                testID="innerView"
            >
                {chartDimensions ? (
                    <VictoryPie
                        padding={aggregateGridPadding(style.slices)}
                        data={series.map((_series, index) => ({
                            ..._series,
                            labelStyle: {
                                fill: normalizedSliceColors[index],
                                ...style.slices?.customStyles?.[_series.stylingKey!]?.label
                            }
                        }))}
                        height={chartDimensions.width}
                        width={chartDimensions.width}
                        style={{
                            labels: { ...createLabelStyleGetters(...LABEL_STYLE_KEYS) }
                        }}
                        labels={({ datum }) => (showLabels ? datum.x : undefined)}
                        // typings are missing `sortOrder`
                        /* eslint-disable-next-line @typescript-eslint/ban-ts-comment */
                        // @ts-ignore
                        sortOrder={sortOrder}
                        innerRadius={
                            presentation === "doughnut"
                                ? style.slices?.innerRadius ?? chartDimensions.width / DEFAULT_INNER_RADIUS_RATIO
                                : undefined
                        }
                        colorScale={normalizedSliceColors}
                    />
                ) : null}
            </View>
        </View>
    );
}

function createLabelStyleGetters(...keys: string[]): VictoryStyleObject {
    const result = {};
    for (const key of keys) {
        Object.assign(result, {
            [key]: ({ datum }: CallbackArgs) => datum.labelStyle[key]
        });
    }

    return result;
}

function aggregateGridPadding(style?: SlicesStyle): Option<BlockProps> {
    if (!style) {
        return;
    }

    const { padding, paddingHorizontal, paddingVertical, paddingTop, paddingRight, paddingBottom, paddingLeft } = style;
    return {
        top: paddingTop ?? paddingVertical ?? padding,
        right: paddingRight ?? paddingHorizontal ?? padding,
        bottom: paddingBottom ?? paddingVertical ?? padding,
        left: paddingLeft ?? paddingHorizontal ?? padding
    };
}
