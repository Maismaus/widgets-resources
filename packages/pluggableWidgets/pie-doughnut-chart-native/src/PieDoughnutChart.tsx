import { createElement, ReactElement } from "react";
import { all } from "deepmerge";

import { PieDoughnutChartProps } from "../typings/PieDoughnutChartProps";
import { PieDoughnutChart as PieDoughnutChartComponent } from "./components/PieDoughnutChart";
import { ChartStyle, defaultStyle } from "./ui/Styles";
import { useSeries } from "./utils/SeriesLoader";

export function PieDoughnutChart(props: PieDoughnutChartProps<ChartStyle>): ReactElement | null {
    const { series: chartSeries, style, ...rest } = props;
    const series = useSeries(chartSeries);

    if (!series) {
        return null;
    }

    return (
        <PieDoughnutChartComponent
            series={series.flatMap(_series => _series.slices)}
            style={all<ChartStyle>([defaultStyle, ...style.filter(o => o != null)])}
            {...rest}
        />
    );
}
