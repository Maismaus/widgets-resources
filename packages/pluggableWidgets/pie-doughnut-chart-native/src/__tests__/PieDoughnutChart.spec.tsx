import { EditableValueBuilder, ListAttributeValueBuilder, ListValueBuilder } from "@mendix/piw-utils-internal";
import { act, fireEvent, render, Thenable } from "react-native-testing-library";
import { ReactTestInstance } from "react-test-renderer";
import { ObjectItem } from "mendix";
import { View } from "react-native";
import { createElement } from "react";
import { Big } from "big.js";

import { PieDoughnutChartProps, SeriesType } from "../../typings/PieDoughnutChartProps";
import { ChartStyle, defaultStyle } from "../ui/Styles";
import { PieDoughnutChart } from "../PieDoughnutChart";

jest.mock("victory-native", () => ({
    VictoryPie: jest.fn(({ children, ...rest }) => (
        <View testID="mock" {...rest}>
            {children}
        </View>
    ))
}));

describe("PieDoughnutChart", () => {
    describe("configuration", () => {
        it("basic static props", () => {
            const { getByTestId, toJSON } = render(<PieDoughnutChart {...createProps()} />);
            fireLayoutEvent(getByTestId("innerView"));

            expect(toJSON()).toMatchSnapshot();
        });

        it("renders doughnut correctly with default inner radius", () => {
            const { getByTestId } = render(<PieDoughnutChart {...createProps({ presentation: "doughnut" })} />);
            fireLayoutEvent(getByTestId("innerView"));

            expect(getByTestId("mock").props.innerRadius).toEqual(25);
        });

        it("renders doughnut correctly with custom inner radius", () => {
            const { getByTestId } = render(
                <PieDoughnutChart
                    {...createProps({
                        presentation: "doughnut",
                        style: [
                            {
                                slices: {
                                    innerRadius: 111
                                }
                            }
                        ]
                    })}
                />
            );
            fireLayoutEvent(getByTestId("innerView"));

            expect(getByTestId("mock").props.innerRadius).toEqual(111);
        });

        it("handles labels", () => {
            const { getByTestId } = render(<PieDoughnutChart {...createProps({ showLabels: true })} />);
            fireLayoutEvent(getByTestId("innerView"));

            expect(getByTestId("mock").props.labels({ datum: { x: "label1" } })).toEqual("label1");
        });

        it("handles sortOrder", () => {
            const { getByTestId, toJSON } = render(<PieDoughnutChart {...createProps({ sortOrder: "descending" })} />);
            fireLayoutEvent(getByTestId("innerView"));

            expect(toJSON()).toMatchSnapshot();
        });
    });

    describe("data", () => {
        const style = [
            {
                slices: {
                    colorPalette: "red;green",
                    customStyles: {
                        key1: {
                            slice: {
                                color: "watermelon"
                            },
                            label: {
                                fontSize: 11
                            }
                        }
                    }
                }
            }
        ] as ChartStyle[];

        it("should create correct series data", () => {
            const { getByTestId } = render(
                <PieDoughnutChart {...createProps({ style, series: setupBasicSeries() })} />
            );
            fireLayoutEvent(getByTestId("innerView"));
            const mock = getByTestId("mock");

            expect(mock.props.data).toEqual([
                { labelStyle: { fill: "watermelon", fontSize: 11 }, stylingKey: "key1", x: "Formatted one", y: 3 },
                { labelStyle: { fill: "red" }, stylingKey: undefined, x: "Formatted two", y: 6 }
            ]);
        });

        it("should create correct label style getters", () => {
            const { getByTestId } = render(
                <PieDoughnutChart {...createProps({ style, series: setupBasicSeries() })} />
            );
            fireLayoutEvent(getByTestId("innerView"));
            const mock = getByTestId("mock");

            expect(mock.props.style.labels.fill({ datum: { labelStyle: { fill: "red" } } })).toEqual("red");
            expect(mock.props.style.labels.fontFamily({ datum: { labelStyle: { fontFamily: "Helvetica" } } })).toEqual(
                "Helvetica"
            );
            expect(mock.props.style.labels.fontSize({ datum: { labelStyle: { fontSize: 123 } } })).toEqual(123);
            expect(mock.props.style.labels.fontStyle({ datum: { labelStyle: { fontStyle: "italic" } } })).toEqual(
                "italic"
            );
            expect(mock.props.style.labels.fontWeight({ datum: { labelStyle: { fontWeight: "200" } } })).toEqual("200");
        });

        it("should create correct colorScale", () => {
            const style = [
                {
                    slices: {
                        colorPalette: "red;green;purple;yellow;pink;blue",
                        customStyles: {
                            key1: {
                                slice: {
                                    color: "watermelon"
                                }
                            },
                            key2: {
                                slice: {
                                    color: "dullgrey"
                                }
                            },
                            key4: {
                                slice: {
                                    color: "brightred"
                                }
                            }
                        }
                    }
                }
            ] as ChartStyle[];

            const { getByTestId } = render(
                <PieDoughnutChart {...createProps({ style, series: setupComplexSeries() })} />
            );
            fireLayoutEvent(getByTestId("innerView"));
            const mock = getByTestId("mock");

            expect(mock.props.colorScale).toEqual(["watermelon", "dullgrey", "red", "brightred", "green"]);
        });
    });

    describe("padding", () => {
        it("prioritises side-specific padding", () => {
            const { getByTestId } = render(
                <PieDoughnutChart
                    {...createProps({
                        style: [
                            {
                                slices: {
                                    paddingTop: 1,
                                    paddingBottom: 2,
                                    paddingLeft: 3,
                                    paddingRight: 4,
                                    paddingHorizontal: 50,
                                    paddingVertical: 51,
                                    padding: 100
                                }
                            }
                        ]
                    })}
                />
            );
            fireLayoutEvent(getByTestId("innerView"));
            const mock = getByTestId("mock");

            expect(mock.props.padding).toEqual({
                top: 1,
                bottom: 2,
                left: 3,
                right: 4
            });
        });

        it("then plane-specific padding", () => {
            const { getByTestId } = render(
                <PieDoughnutChart
                    {...createProps({
                        style: [
                            {
                                slices: {
                                    paddingHorizontal: 50,
                                    paddingVertical: 51,
                                    padding: 100
                                }
                            }
                        ]
                    })}
                />
            );
            fireLayoutEvent(getByTestId("innerView"));
            const mock = getByTestId("mock");

            expect(mock.props.padding).toEqual({
                top: 51,
                bottom: 51,
                left: 50,
                right: 50
            });
        });

        it("then falls-back to padding", () => {
            const { getByTestId } = render(
                <PieDoughnutChart
                    {...createProps({
                        style: [
                            {
                                slices: {
                                    paddingVertical: 51,
                                    padding: 100
                                }
                            }
                        ]
                    })}
                />
            );
            fireLayoutEvent(getByTestId("innerView"));
            const mock = getByTestId("mock");

            expect(mock.props.padding).toEqual({
                top: 51,
                bottom: 51,
                left: 100,
                right: 100
            });
        });
    });
});

function createProps(opts: Partial<PieDoughnutChartProps<ChartStyle>> = {}): PieDoughnutChartProps<ChartStyle> {
    const { series = [], style = [], presentation = "pie", sortOrder = "ascending", showLabels = false } = opts;

    return {
        name: "PieDoughnutTestWidget",
        showLabels,
        sortOrder,
        presentation,
        series,
        style: [defaultStyle, ...style]
    };
}

const fireLayoutEvent = (view: ReactTestInstance): Thenable =>
    act(() => {
        fireEvent(view, "layout", {
            nativeEvent: {
                layout: {
                    width: 300
                }
            }
        });
    });

function setupBasicSeries(): SeriesType[] {
    const xAttribute = new ListAttributeValueBuilder<string>().build();
    const getXAttributeMock = jest.fn();
    getXAttributeMock.mockReturnValueOnce(new EditableValueBuilder<string>().withValue("one").build());
    getXAttributeMock.mockReturnValueOnce(new EditableValueBuilder<string>().withValue("two").build());
    xAttribute.get = getXAttributeMock;

    const yAttribute = new ListAttributeValueBuilder<Big>().build();
    const getYAttributeMock = jest.fn();
    getYAttributeMock.mockReturnValueOnce(new EditableValueBuilder<Big>().withValue(new Big(3)).build());
    getYAttributeMock.mockReturnValueOnce(new EditableValueBuilder<Big>().withValue(new Big(6)).build());
    yAttribute.get = getYAttributeMock;

    const sliceStylingKey = new ListAttributeValueBuilder<string>().build();
    const getStylingKeyMock = jest.fn();
    getStylingKeyMock.mockReturnValueOnce(new EditableValueBuilder<string>().withValue("key1").build());
    getStylingKeyMock.mockReturnValue(new EditableValueBuilder<string>().build());
    sliceStylingKey.get = getStylingKeyMock;

    return [
        {
            dataSource: ListValueBuilder().simple(),
            XAttribute: xAttribute,
            YAttribute: yAttribute,
            sliceStylingKey
        }
    ];
}

function setupComplexSeries(): SeriesType[] {
    const xAttribute = new ListAttributeValueBuilder<string>().build();
    const getXAttributeMock = jest.fn();
    getXAttributeMock.mockReturnValueOnce(new EditableValueBuilder<string>().withValue("one").build());
    getXAttributeMock.mockReturnValueOnce(new EditableValueBuilder<string>().withValue("two").build());
    getXAttributeMock.mockReturnValueOnce(new EditableValueBuilder<string>().withValue("three").build());
    getXAttributeMock.mockReturnValueOnce(new EditableValueBuilder<string>().withValue("four").build());
    getXAttributeMock.mockReturnValueOnce(new EditableValueBuilder<string>().withValue("five").build());
    xAttribute.get = getXAttributeMock;

    const yAttribute = new ListAttributeValueBuilder<Big>().build();
    const getYAttributeMock = jest.fn();
    getYAttributeMock.mockReturnValueOnce(new EditableValueBuilder<Big>().withValue(new Big(3)).build());
    getYAttributeMock.mockReturnValueOnce(new EditableValueBuilder<Big>().withValue(new Big(6)).build());
    getYAttributeMock.mockReturnValueOnce(new EditableValueBuilder<Big>().withValue(new Big(7)).build());
    getYAttributeMock.mockReturnValueOnce(new EditableValueBuilder<Big>().withValue(new Big(12)).build());
    getYAttributeMock.mockReturnValueOnce(new EditableValueBuilder<Big>().withValue(new Big(15)).build());
    yAttribute.get = getYAttributeMock;

    const sliceStylingKey = new ListAttributeValueBuilder<string>().build();
    const getStylingKeyMock = jest.fn();
    getStylingKeyMock.mockReturnValueOnce(new EditableValueBuilder<string>().withValue("key1").build());
    getStylingKeyMock.mockReturnValueOnce(new EditableValueBuilder<string>().withValue("key2").build());
    getStylingKeyMock.mockReturnValueOnce(new EditableValueBuilder<string>().build());
    getStylingKeyMock.mockReturnValueOnce(new EditableValueBuilder<string>().withValue("key4").build());
    getStylingKeyMock.mockReturnValueOnce(new EditableValueBuilder<string>().build());
    sliceStylingKey.get = getStylingKeyMock;

    return [
        {
            dataSource: ListValueBuilder().withItems([
                { id: "1" },
                { id: "2" },
                { id: "3" },
                { id: "4" },
                { id: "5" }
            ] as ObjectItem[]),
            XAttribute: xAttribute,
            YAttribute: yAttribute,
            sliceStylingKey
        }
    ];
}
