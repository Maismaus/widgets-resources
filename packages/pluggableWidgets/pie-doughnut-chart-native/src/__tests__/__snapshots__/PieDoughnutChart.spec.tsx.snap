// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PieDoughnutChart configuration basic static props 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
    testID="innerView"
  >
    <View
      colorScale={Array []}
      data={Array []}
      height={300}
      labels={[Function]}
      padding={
        Object {
          "bottom": 0,
          "left": 0,
          "right": 0,
          "top": 0,
        }
      }
      sortOrder="ascending"
      style={
        Object {
          "labels": Object {
            "fill": [Function],
            "fontFamily": [Function],
            "fontSize": [Function],
            "fontStyle": [Function],
            "fontWeight": [Function],
          },
        }
      }
      testID="mock"
      width={300}
    />
  </View>
</View>
`;

exports[`PieDoughnutChart configuration handles sortOrder 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
    testID="innerView"
  >
    <View
      colorScale={Array []}
      data={Array []}
      height={300}
      labels={[Function]}
      padding={
        Object {
          "bottom": 0,
          "left": 0,
          "right": 0,
          "top": 0,
        }
      }
      sortOrder="descending"
      style={
        Object {
          "labels": Object {
            "fill": [Function],
            "fontFamily": [Function],
            "fontSize": [Function],
            "fontStyle": [Function],
            "fontWeight": [Function],
          },
        }
      }
      testID="mock"
      width={300}
    />
  </View>
</View>
`;
